File: lib/python/opentoken/src/main/opentoken/io/csv/person_attributes_csv_reader.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import csv
import logging
from typing import Dict, Type, Set
from opentoken.io.person_attributes_reader import PersonAttributesReader
from opentoken.attributes.attribute import Attribute
from opentoken.attributes.attribute_loader import AttributeLoader


logger = logging.getLogger(__name__)


class PersonAttributesCSVReader(PersonAttributesReader):
    """
    Reads person attributes from a CSV file.
    Implements the PersonAttributesReader interface.
    """

    def __init__(self, file_path: str):
        """
        Initialize the class with the input file in CSV format.

        Args:
            file_path: The input file path.

        Raises:
            IOError: If an I/O error occurs.
        """
        try:
            self.file_path = file_path
            self.file_handle = open(file_path, 'r', encoding='utf-8')
            self.csv_reader = csv.DictReader(self.file_handle)
            self.iterator = iter(self.csv_reader)
            self.attribute_map: Dict[str, Attribute] = {}

            # Load attributes and build the mapping
            attributes: Set[Attribute] = AttributeLoader.load()
            for header_name in self.csv_reader.fieldnames or []:
                for attribute in attributes:
                    for alias in attribute.get_aliases():
                        if header_name.lower() == alias.lower():
                            self.attribute_map[header_name] = attribute
                            break

        except IOError as e:
            logger.error(f"Error in reading CSV file: {e}")
            raise

    def __iter__(self):
        """Return the iterator object."""
        return self

    def __next__(self) -> Dict[Type[Attribute], str]:
        """
        Get the next record from the CSV file.

        Returns:
            A person attributes map.
        """
        record = next(self.iterator)

        person_attributes: Dict[Type[Attribute], str] = {}
        for key, value in record.items():
            attribute = self.attribute_map.get(key)
            if attribute is not None:
                person_attributes[type(attribute)] = value
            # else ignore attribute as it's not supported

        return person_attributes

    def close(self) -> None:
        """Close the CSV reader and file handle."""
        if self.file_handle:
            self.file_handle.close()

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/io/csv/PersonAttributesCSVReader.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import java.io.IOException;
import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Set;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.truveta.opentoken.io.PersonAttributesReader;
import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.AttributeLoader;

/**
 * Reads person attributes from a CSV file.
 * Implements the {@link PersonAttributesReader} interface.
 */
public class PersonAttributesCSVReader implements PersonAttributesReader {
    private static final Logger logger = LoggerFactory.getLogger(PersonAttributesCSVReader.class);

    private final Reader reader;
    private final CSVParser csvParser;
    private Iterator<CSVRecord> iterator;
    private Map<String, Attribute> attributeMap = new HashMap<>();

    /**
     * Initialize the class with the input file in CSV format.
     * 
     * @param filePath the input file path
     * @throws IOException if an I/O error occurs
     */
    public PersonAttributesCSVReader(String filePath) throws IOException {
        try {

            reader = Files.newBufferedReader(Paths.get(filePath));
            csvParser = new CSVParser(reader, CSVFormat.Builder.create().setHeader().build());
            iterator = csvParser.iterator();

            Set<Attribute> attributes = AttributeLoader.load();
            for (String headerName : csvParser.getHeaderNames()) {
                for (Attribute attribute : attributes) {
                    for (String alias : attribute.getAliases()) {
                        if (headerName.equalsIgnoreCase(alias)) {
                            attributeMap.put(headerName, attribute);
                        }
                    }
                }
            }
        } catch (IOException e) {
            logger.error("Error in reading CSV file: {}", e.getMessage());
            throw e;
        }
    }

    @Override
    public boolean hasNext() {
        return iterator.hasNext();
    }

    @Override
    public void close() throws Exception {
        csvParser.close();
        reader.close();
    }

    @Override
    public Map<Class<? extends Attribute>, String> next() {
        CSVRecord record = iterator.next();

        Map<Class<? extends Attribute>, String> personAttributes = new HashMap<>();
        record.toMap().forEach((key, value) -> {
            Attribute attribute = attributeMap.get(key);
            if (attribute != null) {
                personAttributes.put(attribute.getClass(), value);
            }
            // else ignore attribute as it's not supported
        });

        return personAttributes;
    }
}